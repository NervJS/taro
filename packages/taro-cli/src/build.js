const fs = require('fs-extra')
const path = require('path')
const chalk = require('chalk')
const _ = require('lodash')

const Util = require('./util')
const CONFIG = require('./config')

const appPath = process.cwd()


function build (args, buildConfig) {
  const { type, watch } = buildConfig
  const configDir = require(path.join(appPath, Util.PROJECT_CONFIG))(_.merge)
  const outputPath = path.join(appPath, configDir.outputRoot || CONFIG.OUTPUT_DIR)
  if (!fs.existsSync(outputPath)) {
    fs.mkdirSync(outputPath)
  } else {
    if (type !== Util.BUILD_TYPES.H5) {
      Util.emptyDirectory(outputPath)
    }
  }
  switch (type) {
    case Util.BUILD_TYPES.H5:
      buildForH5({ watch })
      break
    case Util.BUILD_TYPES.WEAPP:
      buildForWeapp({ watch })
      break
    case Util.BUILD_TYPES.SWAN:
      buildForSwan({ watch })
      break
    case Util.BUILD_TYPES.ALIPAY:
      buildForAlipay({ watch })
      break
    case Util.BUILD_TYPES.TT:
      buildForTt({ watch })
      break
    case Util.BUILD_TYPES.RN:
      buildForRN({ watch })
      break
    case Util.BUILD_TYPES.UI:
      buildForUILibrary({ watch })
      break
    default:
      console.log(chalk.red('输入类型错误，目前只支持 weapp/h5/rn/swan/alipay/tt 六端类型'))
  }
}

function buildForWeapp ({ watch }) {
  require('./weapp').build({
    watch,
    adapter: Util.BUILD_TYPES.WEAPP
  })
}

function buildForSwan ({ watch }) {
  require('./weapp').build({
    watch,
    adapter: Util.BUILD_TYPES.SWAN
  })
}

function buildForAlipay ({ watch }) {
  require('./weapp').build({
    watch,
    adapter: Util.BUILD_TYPES.ALIPAY
  })
}

function buildForTt ({ watch }) {
  require('./weapp').build({
    watch,
    adapter: Util.BUILD_TYPES.TT
  })
}

function buildForH5 (buildConfig) {
  require('./h5').build(buildConfig)
}

function buildForRN ({ watch }) {
  require('./rn').build({ watch })
}

function buildForUILibrary ({ watch }) {
  require('./ui').build({ watch })
}

module.exports = build
