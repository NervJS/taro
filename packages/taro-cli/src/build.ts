import * as path from 'path'
import * as fs from 'fs-extra'
import chalk from 'chalk'
import * as _ from 'lodash'

import * as Util from './util'
import CONFIG from './config'
import { BUILD_TYPES, PROJECT_CONFIG } from './util/constants'
import { IBuildConfig } from './util/types'

export default function build (appPath, buildConfig: IBuildConfig) {
  const { type, watch, platform, port, release } = buildConfig
  const configDir = require(path.join(appPath, PROJECT_CONFIG))(_.merge)
  const outputPath = path.join(appPath, configDir.outputRoot || CONFIG.OUTPUT_DIR)
  if (!fs.existsSync(outputPath)) {
    fs.ensureDirSync(outputPath)
  } else if (type !== BUILD_TYPES.H5 && (type !== BUILD_TYPES.QUICKAPP || !watch)) {
    Util.emptyDirectory(outputPath)
  }
  switch (type) {
    case BUILD_TYPES.H5:
      buildForH5(appPath, { watch, port })
      break
    case BUILD_TYPES.WEAPP:
      buildForWeapp(appPath, { watch })
      break
    case BUILD_TYPES.SWAN:
      buildForSwan(appPath, { watch })
      break
    case BUILD_TYPES.ALIPAY:
      buildForAlipay(appPath, { watch })
      break
    case BUILD_TYPES.TT:
      buildForTt(appPath, { watch })
      break
    case BUILD_TYPES.RN:
      buildForRN(appPath, { watch })
      break
    case BUILD_TYPES.QUICKAPP:
      buildForQuickApp(appPath, { watch, port, release })
      break
    case BUILD_TYPES.QQ:
      buildForQQ(appPath, { watch })
      break
    case BUILD_TYPES.UI:
      buildForUILibrary(appPath, { watch })
      break
    case BUILD_TYPES.PLUGIN:
      buildForPlugin(appPath, {
        watch,
        platform
      })
      break
    default:
      console.log(chalk.red('输入类型错误，目前只支持 weapp/swan/alipay/tt/h5/quickapp/rn 七端类型'))
  }
}

function buildForWeapp (appPath: string, { watch }: IBuildConfig) {
  require('./mini').build(appPath, {
    watch,
    adapter: BUILD_TYPES.WEAPP
  })
}

function buildForSwan (appPath: string, { watch }: IBuildConfig) {
  require('./mini').build(appPath, {
    watch,
    adapter: BUILD_TYPES.SWAN
  })
}

function buildForAlipay (appPath: string, { watch }: IBuildConfig) {
  require('./mini').build(appPath, {
    watch,
    adapter: BUILD_TYPES.ALIPAY
  })
}

function buildForTt (appPath: string, { watch }: IBuildConfig) {
  require('./mini').build(appPath, {
    watch,
    adapter: BUILD_TYPES.TT
  })
}

function buildForH5 (appPath: string, buildConfig: IBuildConfig) {
  require('./h5').build(appPath, buildConfig)
}

function buildForRN (appPath: string, { watch }: IBuildConfig) {
  require('./rn').build(appPath, { watch })
}

function buildForQuickApp (appPath: string, { watch, port, release }: IBuildConfig) {
  require('./mini').build(appPath, {
    watch,
    adapter: BUILD_TYPES.QUICKAPP,
    port,
    release
  })
}

function buildForQQ (appPath: string, { watch }: IBuildConfig) {
  require('./mini').build(appPath, {
    watch,
    adapter: BUILD_TYPES.QQ
  })
}

function buildForUILibrary (appPath: string, { watch }: IBuildConfig) {
  require('./ui').build(appPath, { watch })
}

function buildForPlugin (appPath: string, { watch, platform }) {
  const typeMap = {
    [BUILD_TYPES.WEAPP]: '微信',
    [BUILD_TYPES.ALIPAY]: '支付宝'
  }
  if (platform !== BUILD_TYPES.WEAPP && platform !== BUILD_TYPES.ALIPAY) {
    console.log(chalk.red('目前插件编译仅支持 微信/支付宝 小程序！'))
    return
  }
  console.log(chalk.green(`开始编译${typeMap[platform]}小程序插件`))
  require('./plugin').build(appPath, { watch, platform })
}
