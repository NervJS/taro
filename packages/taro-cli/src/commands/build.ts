import { Kernel } from '@tarojs/service'

export default function build (kernel: Kernel, {
  platform,
  isWatch,
  release,
  port,
  ui,
  uiIndex,
  page,
  component,
  envHasBeenSet = false,
  plugin,
  isHelp
}: {
  platform: string,
  isWatch: boolean,
  release?: boolean
  port?: number
  ui?: boolean
  uiIndex?: string
  page?: string
  component?: string
  envHasBeenSet?: boolean
  plugin?: string | boolean
  isHelp?: boolean
}) {
  if (plugin) {
    if (typeof plugin === 'boolean') {
      plugin = 'weapp'
    }
    platform = 'plugin'
  }
  if (platform === 'plugin') {
    plugin = plugin || 'weapp'
  }
  if (ui) {
    platform = 'ui'
  }
  let nodeEnv = process.env.NODE_ENV
  if (!nodeEnv) {
    if (isWatch) {
      nodeEnv = 'development'
    } else {
      nodeEnv = 'production'
    }
  }
  process.env.NODE_ENV = nodeEnv
  process.env.TARO_ENV = platform
  kernel.run({
    name: 'build',
    opts: {
      platform,
      isWatch,
      release,
      port,
      ui,
      uiIndex,
      page,
      component,
      envHasBeenSet,
      plugin,
      isHelp
    }
  })
}
