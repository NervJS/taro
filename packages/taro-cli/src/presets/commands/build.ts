import { IPluginContext } from '@tarojs/service'
import * as hooks from '../constant'
import configValidator from '../../doctor/configValidator'

export default (ctx: IPluginContext) => {
  registerBuildHooks(ctx)
  ctx.registerCommand({
    name: 'build',
    optionsMap: {
      '--type [typeName]': 'Build type, weapp/swan/alipay/tt/qq/jd/h5',
      '--watch': 'Watch mode',
      '--env [env]': 'Value for process.env.NODE_ENV',
      '--blended': 'Blended Taro project in an original MiniApp project'
      // '--port [port]': 'Specified port',
    },
    synopsisList: [
      'taro build --type weapp',
      'taro build --type weapp --watch',
      'taro build --type weapp --env production',
      'taro build --type weapp --blended',
      'taro build native-components --type weapp'
    ],
    async fn (opts) {
      const { options, config, _ } = opts
      const { platform, isWatch, blended } = options
      const { fs, chalk, PROJECT_CONFIG } = ctx.helper
      const { outputPath, configPath } = ctx.paths

      if (!configPath || !fs.existsSync(configPath)) {
        console.log(chalk.red(`找不到项目配置文件${PROJECT_CONFIG}，请确定当前目录是 Taro 项目根目录!`))
        process.exit(1)
      }

      if (typeof platform !== 'string') {
        console.log(chalk.red('请传入正确的编译类型！'))
        process.exit(0)
      }

      // 校验 Taro 项目配置
      const checkResult = await checkConfig({
        configPath,
        projectConfig: ctx.initialConfig
      })
      if (checkResult.lines.length) {
        const NOTE_VALID = chalk.yellow('[!] ')
        const NOTE_INVALID = chalk.red('[✗] ')

        const lineChalk = chalk.hex('#fff')
        const errorChalk = chalk.hex('#f00')
        console.log(errorChalk(`Taro 配置有误，请检查！ (${configPath})`))
        checkResult.lines.forEach(line => {
          console.log(
            '  ' +
            (line.valid ? NOTE_VALID : NOTE_INVALID) +
            lineChalk(line.desc)
          )
        })
        process.exit(0)
      }

      const isProduction = process.env.NODE_ENV === 'production' || !isWatch

      // dist folder
      fs.ensureDirSync(outputPath)

      // is build native components mode?
      const isBuildNativeComp = _[1] === 'native-components'

      await ctx.applyPlugins(hooks.ON_BUILD_START)
      await ctx.applyPlugins({
        name: platform,
        opts: {
          config: {
            ...config,
            isWatch,
            mode: isProduction ? 'production' : 'development',
            blended,
            isBuildNativeComp,
            async modifyWebpackChain (chain, webpack) {
              await ctx.applyPlugins({
                name: hooks.MODIFY_WEBPACK_CHAIN,
                initialVal: chain,
                opts: {
                  chain,
                  webpack
                }
              })
            },
            async modifyBuildAssets (assets, miniPlugin) {
              await ctx.applyPlugins({
                name: hooks.MODIFY_BUILD_ASSETS,
                initialVal: assets,
                opts: {
                  assets,
                  miniPlugin
                }
              })
            },
            async modifyMiniConfigs (configMap) {
              await ctx.applyPlugins({
                name: hooks.MODIFY_MINI_CONFIGS,
                initialVal: configMap,
                opts: {
                  configMap
                }
              })
            },
            async onCompilerMake (compilation) {
              await ctx.applyPlugins({
                name: hooks.ON_COMPILER_MAKE,
                opts: {
                  compilation
                }
              })
            },
            async onBuildFinish ({ error, stats, isWatch }) {
              await ctx.applyPlugins({
                name: hooks.ON_BUILD_FINISH,
                opts: {
                  error,
                  stats,
                  isWatch
                }
              })
            }
          }
        }
      })
    }
  })
}

function registerBuildHooks (ctx) {
  [
    hooks.MODIFY_WEBPACK_CHAIN,
    hooks.MODIFY_BUILD_ASSETS,
    hooks.MODIFY_MINI_CONFIGS,
    hooks.ON_COMPILER_MAKE,
    hooks.ON_BUILD_START,
    hooks.ON_BUILD_FINISH
  ].forEach(methodName => {
    ctx.registerMethod(methodName)
  })
}

async function checkConfig ({ projectConfig, configPath }) {
  const result = await configValidator({
    configPath,
    projectConfig
  })
  return result
}
