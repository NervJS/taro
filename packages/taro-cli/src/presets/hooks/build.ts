import { IPluginContext } from '@tarojs/service'

import * as hooks from '../constant'

export default (ctx: IPluginContext) => {
  [
    hooks.MODIFY_WEBPACK_CHAIN,
    hooks.MODIFY_BUILD_ASSETS,
    hooks.MODIFY_MINI_CONFIGS,
    hooks.MODIFY_COMPONENT_CONFIG,
    hooks.ON_COMPILER_MAKE,
    hooks.ON_PARSE_CREATE_ELEMENT,
    hooks.ON_BUILD_START,
    hooks.ON_BUILD_FINISH,
    hooks.ON_BUILD_COMPLETE,
    hooks.MODIFY_RUNNER_OPTS
  ].forEach(methodName => {
    ctx.registerMethod(methodName)
  })
}
