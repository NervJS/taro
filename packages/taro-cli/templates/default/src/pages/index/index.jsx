<%if (['react', 'preact'].includes(framework)) {-%>
import { Component<% if (typescript) {%>, PropsWithChildren<%}%> } from 'react'
<%}-%>
import { View, Text } from '@tarojs/components'
import './index.<%= cssExt %>'

export default class <%= _.capitalize(pageName) %> extends <% if (typescript) {%>Component<PropsWithChildren><%} else {%>Component<%}%> {

  componentWillMount () { }

  componentDidMount () { }

  componentWillUnmount () { }

  componentDidShow () { }

  componentDidHide () { }

  render () {
    return (
      <View className='<%= pageName %>'>
        <Text>Hello world!</Text>
      </View>
    )
  }
}
