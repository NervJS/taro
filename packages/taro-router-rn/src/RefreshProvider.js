import React from 'react'
import { ScrollView, RefreshControl } from 'react-native'

class RefreshProvider extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      refreshing: false
    }
  }

  render () {
    const {enablePullDownRefresh, disableScroll} = this.props
    if (disableScroll) {
      return this.props.children
    } else {
      return (
        <ScrollView
          style={{flex: 1}}
          scrollEventThrottle={5}
          alwaysBounceVertical={false}
          onScroll={this.onScroll}
          refreshControl={
            enablePullDownRefresh
              ? <RefreshControl
                refreshing={this.state.refreshing}
                onRefresh={this.handlePullDownRefresh}
              />
              : null
          }
        >
          {this.props.children && this.props.children}
        </ScrollView>
      )
    }
  }

  onScroll = (e) => {
    this.props.onScroll && this.props.onScroll(e.nativeEvent.contentOffset.y)
  }

  // TODO 滚动到底部事件
  handleReachBottom = () => {
    this.props.onReachBottom && this.props.onReachBottom()
  }

  handlePullDownRefresh = (callback) => {
    this.setState({refreshing: true}, callback)
    try {
      // TODO 处理异步的情况
      this.props.onPullDownRefresh && this.props.onPullDownRefresh()
    } catch (e) {
      throw new Error(e)
    } finally {
      this.setState({refreshing: false})
    }
  }

  stopPullDownRefresh = (callback) => {
    this.setState({refreshing: false}, callback)
  }
}

export default RefreshProvider
