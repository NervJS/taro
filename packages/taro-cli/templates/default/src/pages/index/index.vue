<template>
  <view class="<%= pageName %>">
    <text>{{ msg }}</text>
  </view>
</template>

<script>
<%if (framework === 'vue3') {-%>
import { ref } from 'vue'
<%}-%>
import './index.<%= cssExt %>'

export default {
<%if (framework === 'vue') {-%>
  data () {
    return {
      msg: 'Hello world!'
    }
  }
<%}-%>
<%if (framework === 'vue3') {-%>
  setup () {
    const msg = ref('Hello world')
    return {
      msg
    }
  }
<%}-%>
}
</script>
