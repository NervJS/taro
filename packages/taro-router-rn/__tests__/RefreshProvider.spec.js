import { shallow } from 'enzyme/build/index'
import toJson from 'enzyme-to-json'
import React from 'react'
import RefreshProvider from '../src/RefreshProvider'

class HomeScreen extends React.Component {
  constructor (props) {
    super(props)
  }

  render () {
    return <View>Test</View>
  }
}

describe('RefreshProvider', function () {
  let Taro = {}
  it('should render success', function () {
    const wrapper = shallow(<RefreshProvider Taro={Taro}><HomeScreen /></RefreshProvider>)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should call onScroll success', function () {
    const mockCallback = jest.fn()
    const wrapper = shallow(<RefreshProvider Taro={Taro}><HomeScreen /></RefreshProvider>)
    wrapper.setProps({onScroll: mockCallback})
    wrapper.instance().onScroll({nativeEvent: {contentOffset: {}}})
    expect(mockCallback.mock.calls.length).toBe(1)
  })

  it('should call handleReachBottom success', function () {
    const mockCallback = jest.fn()
    const wrapper = shallow(<RefreshProvider Taro={Taro}><HomeScreen /></RefreshProvider>)
    wrapper.setProps({onReachBottom: mockCallback})
    wrapper.instance().handleReachBottom()
    expect(mockCallback.mock.calls.length).toBe(1)
  })

  it('should call handlePullDownRefresh success', function (done) {
    const mockCallback = jest.fn()
    const wrapper = shallow(<RefreshProvider Taro={Taro}><HomeScreen /></RefreshProvider>)
    wrapper.setProps({onPullDownRefresh: mockCallback})

    function callback (data) {
      expect(wrapper.state().refreshing).toBe(true)
      done()
    }

    wrapper.instance().handlePullDownRefresh(callback)
    expect(mockCallback.mock.calls.length).toBe(1)
  })

  it('should call stopPullDownRefresh success', function (done) {
    const wrapper = shallow(<RefreshProvider Taro={Taro}><HomeScreen /></RefreshProvider>)

    function callback (data) {
      expect(wrapper.state().refreshing).toBe(false)
      done()
    }

    wrapper.setState({refreshing: true})
    wrapper.instance().stopPullDownRefresh(callback)
  })
})
