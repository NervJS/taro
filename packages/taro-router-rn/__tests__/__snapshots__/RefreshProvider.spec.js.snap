// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefreshProvider should render success 1`] = `
<ScrollView
  alwaysBounceVertical={false}
  onScroll={[Function]}
  refreshControl={null}
  scrollEventThrottle={5}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <HomeScreen />
</ScrollView>
`;
